text/0096-contract-metadata.md

```diff
@@ -0,0 +1,206 @@
+
+- `contract_name`: A human-readable name for the contract (e.g. "Casper DEX").
+- `contract_description`: A brief description of the contract’s purpose or functionality.
+- `contract_icon_uri`: A URI (URL) pointing to an icon image for the contract (e.g. a logo).
+- `contract_project_uri`: A URI pointing to the contract’s project website or documentation.
+
+Each field is optional – contracts may provide all, some, or none of these. If a field is not applicable, it simply returns `None`. By keeping these fields on-chain and unchangeable after deployment, users and tools can rely on the metadata as the authentic self-declared identity of the contract, improving transparency and reducing reliance on off-chain sources.
+
+## Motivation
+
+[motivation]: #motivation
+
+- Establish a uniform and simple approach for contract metadata storage. 
+- Enhance contract discoverability and transparency for end-users. 
+- Simplify indexing and display logic for tools like wallets and explorers.
+
+## Specification
+
+[specification]: #specification
+
+Contracts implementing this standard must expose the following entry points:
+
+```rust
+pub trait CEP96ContractMetadata {
+    /// Contract's human-readable name.
+    fn contract_name(&self) -> Option<String>;
+
+    /// Brief description of the contract.
+    fn contract_description(&self) -> Option<String>;
+
+    /// URI pointing to the contract's icon image.
+    fn contract_icon_uri(&self) -> Option<String>;
+
+    /// URI pointing to the project's website or documentation.
+    fn contract_project_uri(&self) -> Option<String>;
+}
+```
+
+Each method returns `Option<String>`, allowing contracts to omit any metadata field.
+
+
+## Storage Specification
+
+[storage-specification]: #storage-specification
+
+To enforce simplicity, we recommend storing each metadata field under a dedicated named URef key in the contract state. Each metadata entry (if provided at installation) is stored once and cannot be altered afterward.
+
+**URef keys**:
+- `contract_name`
+- `contract_description`
+- `contract_icon_uri`
+- `contract_project_uri`
+
+If a field is omitted, its named `URef` is simply not created.
+
+Here are the benefits of **URef-based Metadata Storage**:
+
+- **Simplicity**: Easy for developers and tools to implement; minimal complexity.
+- **Discoverability**: Named keys are easy to query directly via RPC.
+- **Efficiency**: No need to parse or manage additional storage structures; direct access to metadata.
+
+## Example Implementation
+
+[example-implementation]: #example-implementation
+
+Here's a minimal Rust implementation demonstrating this storage pattern:
+
+```rust
+use casper_contract::contract_api::{runtime, storage};
+use casper_types::{CLValue, URef};
+
+const NAME_KEY: &str = "contract_name";
+const DESCRIPTION_KEY: &str = "contract_description";
+const ICON_URI_KEY: &str = "contract_icon_uri";
+const PROJECT_URI_KEY: &str = "contract_project_uri";
+
+#[no_mangle]
+pub extern "C" fn init() {
```

Comment:
```suggestion
pub extern "C" fn call() {
```

We don't have "constructors" in the smart contract engine per se, URef generation should happen at the contract creation time. This example should cover that